components/kyma-environment-broker/internal/broker/instance_create.go

```diff
@@ -52,79 +55,120 @@ func NewProvision(cfg Config, instancesStorage storage.Instances, builderFactory
 // Provision creates a new service instance
 //   PUT /v2/service_instances/{instance_id}
 func (b *ProvisionEndpoint) Provision(ctx context.Context, instanceID string, details domain.ProvisionDetails, asyncAllowed bool) (domain.ProvisionedServiceSpec, error) {
-	b.dumper.Dump("Provision instanceID:", instanceID)
-	b.dumper.Dump("Provision details:", details)
-	b.dumper.Dump("Provision asyncAllowed:", asyncAllowed)
+	// validation of incoming input
+	ersContext, parameters, err := b.validate(details)
+	if err != nil {
+		errMsg := fmt.Sprintf("[instanceID: %s] %s", instanceID, err)
+		return domain.ProvisionedServiceSpec{}, apiresponses.NewFailureResponse(err, http.StatusBadRequest, errMsg)
+	}
 
-	// unmarshall ERS context
-	var ersContext internal.ERSContext
-	err := json.Unmarshal(details.RawContext, &ersContext)
+	provisioningParameters := internal.ProvisioningParameters{
+		PlanID:     details.PlanID,
+		ServiceID:  details.ServiceID,
+		ErsContext: ersContext,
+		Parameters: parameters,
+	}
+
+	// check if operation with instance ID already created
+	existingOperation, errStorage := b.operationsStorage.GetProvisioningOperationByInstanceID(instanceID)
+	switch {
+	case errStorage != nil && !dberr.IsNotFound(errStorage):
+		b.dumper.Dump("cannot get existing operation from storage", errStorage)
+		return domain.ProvisionedServiceSpec{}, errors.New("cannot get existing operation from storage")
+	case existingOperation != nil && !dberr.IsNotFound(errStorage):
+		return b.handleExistingOperation(existingOperation, provisioningParameters)
+	}
+
+	// create and save new operation
+	operation, err := internal.NewProvisioningOperation(instanceID, provisioningParameters)
 	if err != nil {
-		return domain.ProvisionedServiceSpec{}, errors.Wrap(err, "while decoding context")
+		b.dumper.Dump("cannot create new operation: ", err)
+		return domain.ProvisionedServiceSpec{}, errors.New("cannot create new operation")
 	}
-	if ersContext.GlobalAccountID == "" {
-		return domain.ProvisionedServiceSpec{}, errors.New("GlobalAccountID parameter cannot be empty")
+
+	err = b.operationsStorage.InsertProvisioningOperation(operation)
+	if err != nil {
+		b.dumper.Dump("cannot save operations: ", err)
+		return domain.ProvisionedServiceSpec{}, errors.New("cannot save operations")
 	}
-	b.dumper.Dump("ERS context:", ersContext)
+
+	// add new operation to queue
+	b.queue.Add(operation.ID)
+
+	return domain.ProvisionedServiceSpec{
+		IsAsync:       true,
+		OperationData: operation.ID,
+	}, nil
+}
+
+func (b *ProvisionEndpoint) validate(details domain.ProvisionDetails) (internal.ERSContext, internal.ProvisioningParametersDTO, error) {
```

Comment:
it is not only about validating, it also extracts so the name 'validate' does not describe what happens